astropy_healpix/core.py

```diff
@@ -116,12 +117,72 @@ def nside_to_pixel_resolution(nside):
     -------
     resolution : :class:`~astropy.units.Quantity`
         The resolution of the HEALPix pixels
+
+    See also
+    --------
+    pixel_resolution_to_nside
     """
     nside = np.asanyarray(nside, dtype=np.int64)
     _validate_nside(nside)
     return (nside_to_pixel_area(nside) ** 0.5).to(u.arcmin)
 
 
+def pixel_resolution_to_nside(resolution, round='nearest'):
+    """Find closest HEALPix nside for a given angular resolution.
+
+    This function is the inverse of `nside_to_pixel_resolution`,
+    for the default rounding scheme of ``round='nearest'``.
+
+    If you choose ``round='up'``, you'll get HEALPix pixels that
+    have at least the requested resolution (usually a bit better
+    due to rounding).
+
+    Pixel resolution is defined as square root of pixel area.
+
+    Parameters
+    ----------
+    resolution : `~astropy.units.Quantity`
+        Angular resolution
+    round : {'up', 'nearest', 'down'}
+        Which way to round
+
+    Returns
+    -------
+    nside : int
+        The number of pixels on the side of one of the 12 'top-level' HEALPix tiles.
+        Always a power of 2.
+
+    Examples
+    --------
+    >>> from astropy_healpix import pixel_resolution_to_nside
+    >>> pixel_resolution_to_nside('13 arcmin')
+    256
+    >>> pixel_resolution_to_nside('13 arcmin', round='up')
+    512
+    """
+    resolution = Angle(resolution).radian
+    pixel_area = resolution * resolution
+    npix = 4 * math.pi / pixel_area
+    nside = np.sqrt(npix / 12)
+
+    # Now we have to round to the closest ``nside``
+    # Since ``nside`` must be a power of two,
+    # we first compute the corresponding ``level = log2(nside)`
+    # round the level and then go back to nside
+    level = np.log2(nside)
+
+    if round == 'up':
+        level = np.array(level, dtype=int) + 1
+    elif round == 'nearest':
+        level = np.array(level + 0.5, dtype=int)
+    elif round == 'down':
+        level = np.array(level, dtype=int)
```

Comment:
Rather than doing e.g. +1, +0.5, etc., what about using ``np.round(...)``, ``np.floor(...)`` and ``np.ceil(...)``? (then just add ``.astype(int)``. The issue is that rounding up currently won't give the optimal result if level happens to be an integer since it will return a resolution one too high (but ``ceil`` would do the right thing)

Reply:
Done.